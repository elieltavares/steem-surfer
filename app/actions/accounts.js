import { encryptKey } from '../utils/crypto';
import { setItem, removeItem, getByPrefix } from '../helpers/storage';

export const ACCOUNT_ADDED = 'accounts/ACCOUNT_ADDED';
export const ACCOUNT_ADDED_SC = 'accounts/ACCOUNT_SC_ADDED';
export const ACCOUNT_UPDATED_SC = 'accounts/ACCOUNT_UPDATED_SC';
export const ACCOUNT_DELETED = 'accounts/ACCOUNT_DELETED';
export const ACCOUNTS_DELETED = 'accounts/ACCOUNTS_DELETED';

export const addAccount = (username, keys) => (dispatch, getState) => {
  const { global } = getState();

  const { pin } = global;

  // key encryption
  const eKeys = Object.assign(
    {},
    ...Object.keys(keys).map(k => ({
      [k]: keys[k] ? encryptKey(keys[k], pin) : null
    }))
  );

  const accountData = {
    type: 's',
    username,
    keys: eKeys
  };

  setItem(`account_${username}`, accountData);
  // setItem(`active_account`, username);
  dispatch(accountAdded(username, accountData));
};

export const addAccountSc = (
  username,
  accessToken,
  refreshToken,
  expiresIn,
  updated = false
) => (dispatch, getState) => {
  const { global } = getState();

  const { pin } = global;

  const accountData = {
    type: 'sc',
    username,
    accessToken: encryptKey(accessToken, pin),
    refreshToken: encryptKey(refreshToken, pin),
    expiresIn
  };

  setItem(`account_${username}`, accountData);
  if (!updated) {
    dispatch(accountAddedSc(username, accountData));
  } else {
    dispatch(accountUpdatedSc(username, accountData));
  }
};

export const deleteAccount = username => dispatch => {
  removeItem(`account_${username}`);

  dispatch(accountDeleted(username));
};

export const deleteAccounts = () => dispatch => {
  getByPrefix('account_').forEach(ac => {
    removeItem(`account_${ac.username}`);
  });

  dispatch(accountsDeleted());
};

/* action creators */

export const accountAdded = username => ({
  type: ACCOUNT_ADDED,
  payload: { username }
});

export const accountAddedSc = username => ({
  type: ACCOUNT_ADDED_SC,
  payload: { username }
});

export const accountUpdatedSc = (username, accountData) => ({
  type: ACCOUNT_UPDATED_SC,
  payload: { username, accountData }
});

export const accountDeleted = username => ({
  type: ACCOUNT_DELETED,
  payload: { username }
});

export const accountsDeleted = () => ({
  type: ACCOUNTS_DELETED,
  payload: {}
});
