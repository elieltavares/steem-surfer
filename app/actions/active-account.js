/*
eslint-disable camelcase
 */

import { setItem, removeItem } from '../helpers/storage';
import { getAccounts, getAccountRC } from '../backend/steem-client';

import { getPoints } from '../backend/esteem-client';

export const LOGGED_IN = 'active-account/LOGGED_IN';
export const LOGGED_OUT = 'active-account/LOGGED_OUT';
export const UPDATED = 'active-account/UPDATED';

export const logIn = username => (dispatch, getState) => {
  setItem(`active_account`, username);
  dispatch(loggedIn(username));
  update(username, dispatch, getState);
};

export const logOut = () => dispatch => {
  removeItem(`active_account`);
  dispatch(loggedOut());
};

export const updateActiveAccount = () => (dispatch, getState) => {
  const { activeAccount } = getState();
  const { username } = activeAccount;
  update(username, dispatch, getState);
};

const update = (username, dispatch, getState) => {
  getAccounts([username])
    .then(resp => {
      const account = resp[0];

      return getAccountRC(username).then(r => {
        if (r.rc_accounts && r.rc_accounts.length > 0) {
          return Object.assign({}, account, { rcAccount: r.rc_accounts[0] });
        }

        return account;
      });
    })
    // Synchronising user's blog for reblog detection causes big amount of data usage. disabled.
    // See components/elements/EntryReblogBtn.js:43
    .then(account => Object.assign({}, account, { blog: [] }))
    .catch(() => {})
    .then(account =>
      getPoints(username).then(p => {
        const o = Object.assign({}, account, {
          unclaimed_points: p.unclaimed_points
        });

        // check active user still same
        const { activeAccount: a } = getState();
        const { username: u } = a;

        if (a && u === username) {
          dispatch(updated(username, o));
        }
        return o;
      })
    )
    .catch(() => {});
};

export const loggedIn = username => ({
  type: LOGGED_IN,
  payload: { username }
});

export const loggedOut = () => ({
  type: LOGGED_OUT,
  payload: {}
});

export const updated = (username, accountData) => ({
  type: UPDATED,
  payload: { username, accountData }
});
