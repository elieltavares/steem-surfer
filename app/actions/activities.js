import { getUnreadActivityCount } from '../backend/esteem-client';

export const FETCHED = 'activities/FETCHED';
export const RESET = 'activities/RESET';

export const fetchActivities = username => async (dispatch, getState) => {
  const { activeAccount } = getState();
  if (!activeAccount) {
    return;
  }

  const unread = await getUnreadActivityCount(username);

  dispatch(fetched(unread));
};

export const resetActivities = () => dispatch => {
  dispatch(reset());
};

/* action creators */

export const fetched = unread => ({
  type: FETCHED,
  payload: { unread }
});

export const reset = () => ({
  type: RESET
});
