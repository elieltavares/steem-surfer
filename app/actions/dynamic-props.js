import {
  getDynamicGlobalProperties,
  getFeedHistory,
  getRewardFund
} from '../backend/steem-client';
import parseToken from '../utils/parse-token';

export const FETCHED = 'dynamic-props/FETCHED';

export const fetchGlobalProps = () => async dispatch => {
  let globalDynamic;
  let feedHistory;
  let rewardFund;

  try {
    globalDynamic = await getDynamicGlobalProperties();
    feedHistory = await getFeedHistory();
    rewardFund = await getRewardFund();
  } catch (e) {
    return;
  }

  const steemPerMVests =
    (parseToken(globalDynamic.total_vesting_fund_steem) /
      parseToken(globalDynamic.total_vesting_shares)) *
    1e6;
  const base = parseToken(feedHistory.current_median_history.base);
  const quote = parseToken(feedHistory.current_median_history.quote);
  const fundRecentClaims = rewardFund.recent_claims;
  const fundRewardBalance = parseToken(rewardFund.reward_balance);

  dispatch(fetched(steemPerMVests, base, quote, fundRecentClaims, fundRewardBalance));
};

/* action creators */

export const fetched = (
  steemPerMVests,
  base,
  quote,
  fundRecentClaims,
  fundRewardBalance
) => ({
  type: FETCHED,
  payload: { steemPerMVests, base, quote, fundRecentClaims, fundRewardBalance }
});
