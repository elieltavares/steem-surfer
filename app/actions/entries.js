import {
  getDiscussions,
  getRepliesByLastUpdate
} from '../backend/steem-client';

export const FETCH_BEGIN = 'entries/FETCH_BEGIN';
export const FETCH_OK = 'entries/FETCH_OK';
export const FETCH_ERROR = 'entries/FETCH_ERROR';
export const INVALIDATE = 'entries/INVALIDATE';
export const SET_READ = 'entries/SET_READ';
export const SET_VOTED = 'entries/SET_VOTED';
export const UPDATE_ENTRY = 'entries/UPDATE';

export const makeGroupKeyForEntries = (what, tag = '') => {
  if (tag) {
    return `${what}-${tag}`;
  }
  return `${what}`;
};

export function fetchEntries(what, tag = '', more = false) {
  return (dispatch, getState) => {
    const { entries } = getState();
    const pageSize = 20;

    const groupKey = makeGroupKeyForEntries(what, tag);

    if (!more && entries.getIn([groupKey, 'entries']).size) {
      return;
    }

    const lastEntry = entries
      .getIn([groupKey, 'entries'])
      .valueSeq()
      .last();

    let fn;
    let query;

    if (tag.startsWith('@')) {
      const username = tag.replace('@', '');
      switch (what) {
        case 'feed':
          fn = getDiscussions;
          query = {
            tag: username,
            limit: pageSize,
            start_author: undefined,
            start_permlink: undefined
          };
          break;
        case 'comments':
          fn = getDiscussions;
          query = {
            limit: pageSize,
            start_author: username,
            start_permlink: undefined
          };
          break;
        case 'replies':
          fn = getRepliesByLastUpdate;
          query = {
            start_author: username,
            start_permlink: undefined,
            limit: pageSize
          };
          break;
        default:
          fn = getDiscussions;
          query = {
            tag: username,
            limit: pageSize,
            start_author: undefined,
            start_permlink: undefined
          };
          break;
      }
    } else {
      fn = getDiscussions;
      // make sure tag is not null or undefined. it should be empty string.
      query = {
        tag: tag || '',
        limit: pageSize,
        start_author: undefined,
        start_permlink: undefined
      };
    }

    if (lastEntry) {
      query.start_author = lastEntry.author;
      query.start_permlink = lastEntry.permlink;
    }

    dispatch({
      type: FETCH_BEGIN,
      payload: { group: groupKey }
    });

    const fnArgs = what === 'replies' ? [query] : [what, query];

    fn(...fnArgs)
      .then(resp => {
        dispatch({
          type: FETCH_OK,
          payload: {
            data: resp,
            group: groupKey,
            hasMore: resp.length >= pageSize
          }
        });

        return resp;
      })
      .catch(e => {
        dispatch({
          type: FETCH_ERROR,
          payload: { group: groupKey, error: e }
        });
      });
  };
}

export function updateEntry(newData) {
  return dispatch => {
    dispatch({
      type: UPDATE_ENTRY,
      payload: { data: newData }
    });
  };
}

export function invalidateEntries(what, tag = '') {
  return dispatch => {
    const groupKey = makeGroupKeyForEntries(what, tag);

    dispatch({
      type: INVALIDATE,
      payload: { group: groupKey }
    });
  };
}
