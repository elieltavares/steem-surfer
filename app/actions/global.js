import getSymbolFromCurrency from 'currency-symbol-map';
import { setItem } from '../helpers/storage';
import { setAddress } from '../backend/steem-client';

export const THEME_CHANGED = 'global/THEME_CHANGED';
export const LIST_STYLE_CHANGED = 'global/LIST_STYLE_CHANGED';
export const CURRENCY_CHANGED = 'global/CURRENCY_CHANGED';
export const LOCALE_CHANGED = 'global/LOCALE_CHANGED';
export const PUSH_NOTIFY_CHANGED = 'global/PUSH_NOTIFY_CHANGED';
export const SERVER_CHANGED = 'global/SERVER_CHANGED';

export const PIN_EXPOSED = 'global/PIN_EXPOSED';
export const PIN_WIPED = 'global/PIN_WIPED';

export const SET_INT_CONN = 'global/SET_INT_CONN';

export const changeTheme = () => (dispatch, getState) => {
  const { global } = getState();

  const { theme } = global;
  const newTheme = theme === 'day' ? 'night' : 'day';

  setItem('theme', newTheme);
  window.setTheme(newTheme);

  dispatch(themeChanged(newTheme));
};

export const changeListStyle = () => (dispatch, getState) => {
  const { global } = getState();

  const { listStyle } = global;

  const newStyle = listStyle === 'row' ? 'grid' : 'row';

  setItem('list-style', newStyle);

  dispatch(listStyleChanged(newStyle));
};

export const changeCurrency = (currency, currencyRate) => dispatch => {
  const symbol = getSymbolFromCurrency(currency);

  setItem('currency', currency);
  setItem('currency-rate', currencyRate);
  setItem('currency-symbol', symbol);

  dispatch(currencyChanged(currency, currencyRate, symbol));
};

export const changeLocale = locale => dispatch => {
  setItem('locale', locale);

  dispatch(localeChanged(locale));
};

export const changePushNotify = val => dispatch => {
  setItem('push-notify', val);

  dispatch(pushNotifyChanged(val));
};

export const changeServer = server => dispatch => {
  setItem('server', server);

  setAddress(server);

  dispatch(serverChanged(server));
};

export const exposePin = pin => dispatch => {
  dispatch(pinExposed(pin));
};

export const wipePin = () => dispatch => {
  dispatch(pinWiped());
};

export const setIntConn = val => dispatch => {
  dispatch(intConnSet(val));
};

/* action creators */

export const themeChanged = newTheme => ({
  type: THEME_CHANGED,
  payload: { newTheme }
});

export const listStyleChanged = newStyle => ({
  type: LIST_STYLE_CHANGED,
  payload: { newStyle }
});

export const currencyChanged = (currency, currencyRate, currencySymbol) => ({
  type: CURRENCY_CHANGED,
  payload: { currency, currencyRate, currencySymbol }
});

export const localeChanged = locale => ({
  type: LOCALE_CHANGED,
  payload: { locale }
});

export const pushNotifyChanged = val => ({
  type: PUSH_NOTIFY_CHANGED,
  payload: { val }
});

export const serverChanged = server => ({
  type: SERVER_CHANGED,
  payload: { server }
});

export const pinExposed = pin => ({
  type: PIN_EXPOSED,
  payload: { pin }
});

export const pinWiped = () => ({
  type: PIN_WIPED
});

export const intConnSet = val => ({
  type: SET_INT_CONN,
  payload: { val }
});
