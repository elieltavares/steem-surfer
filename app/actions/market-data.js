import { getMarketData } from '../backend/esteem-client';

export const FETCHED = 'MARKET-DATA/FETCHED';

export const fetchMarketData = () => async dispatch => {
  const data = await getMarketData();

  if (data) {
    dispatch(fetched(data));
  }
};

/* action creators */

export const fetched = data => ({
  type: FETCHED,
  payload: data
});
