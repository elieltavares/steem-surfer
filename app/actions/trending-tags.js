import { Client } from 'dsteem';

export const TT_FETCH_BEGIN = 'TT_FETCH_BEGIN';
export const TT_FETCH_OK = 'TT_FETCH_OK';
export const TT_FETCH_ERROR = 'TT_FETCH_ERROR';

const client = new Client('https://api.steemit.com');

const shuffle = arr => {
  const arry = arr;
  for (let i = arry.length - 1; i > 0; i -= 1) {
    const j = Math.floor(Math.random() * (i + 1));
    const temp = arry[i];
    arry[i] = arry[j];
    arry[j] = temp;
  }
  return arry;
};

export function fetchTrendingTags(afterTag = '', limit = 100) {
  return (dispatch, getState) => {
    const { trendingTags } = getState();

    if (trendingTags.list.length >= 1) {
      return;
    }

    dispatch(fetchBegin());

    client.database
      .call('get_trending_tags', [afterTag, limit])
      .then(resp => {
        const sresp = shuffle(resp);

        dispatch(fetchOk(sresp));

        return sresp;
      })
      .catch(() => {
        dispatch(fetchError());
      });
  };
}

/* action creators */

export const fetchBegin = () => ({
  type: TT_FETCH_BEGIN
});

export const fetchOk = payload => ({
  type: TT_FETCH_OK,
  payload
});

export const fetchError = () => ({
  type: TT_FETCH_ERROR
});
