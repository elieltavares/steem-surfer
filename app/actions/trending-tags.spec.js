/* eslint-disable */
import { spy } from 'sinon';
import * as actions from '../../app/actions/trending-tags';

describe('actions', () => {
  it("should create 'fetch begin' action", () => {
    expect(actions.fetchBegin()).toMatchSnapshot();
  });

  it("should create 'fetch ok' action", () => {
    expect(actions.fetchOk(['life', 'kr', 'steemit', 'art'])).toMatchSnapshot();
  });

  it("should create 'fetch error' action", () => {
    expect(actions.fetchError()).toMatchSnapshot();
  });
});
