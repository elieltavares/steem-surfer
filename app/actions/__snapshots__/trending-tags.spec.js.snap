// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should create 'fetch begin' action 1`] = `
Object {
  "type": "TT_FETCH_BEGIN",
}
`;

exports[`actions should create 'fetch error' action 1`] = `
Object {
  "type": "TT_FETCH_ERROR",
}
`;

exports[`actions should create 'fetch ok' action 1`] = `
Object {
  "payload": Array [
    "life",
    "kr",
    "steemit",
    "art",
  ],
  "type": "TT_FETCH_OK",
}
`;
